src/directives/repeat.ts

```diff
@@ -41,67 +96,139 @@ export function repeat<T>(
+    //   position using the "old key to old index" map.
+    // * If either the new or old pointers move past each other then all we have
+    //   left is additions (if old list exhsusted) or removals (if new list
+    //   exhausted). Those are handled in the final while loops at the end.
+    // * TODO(kschaaf) Note, we could calculate the longest increasing
+    //   subsequence (LIS) of old items in new position, and only move those not
+    //   in the LIS set. However that costs O(nlogn) time and adds a bit more
+    //   code, and only helps make rare types of mutations require fewer moves.
+    //   The above handles removes, adds, reversal, swaps, and single moves of
+    //   contiguous items in linear time, in the minimum number of moves. As
+    //   the number of multiple moves where LIS might help approaches a random
+    //   shuffle, the LIS optimization becomes less helpful, so it seems not
+    //   worth the code at this point, but could consider if a compelling case
+    //   arises.
+    while (oldStartIndex <= oldEndIndex && newStartIndex <= newEndIndex) {
+      if (oldStartPart == null) {
+        // Old part at head has already been used; skip
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (oldEndPart == null) {
+        // Old part at tail has already been used; skip
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newStartResult.key) {
+        // Old head matches new head; update in place
+        newParts[newStartIndex] = updatePart(oldStartPart, newStartResult);
+        oldStartPart = oldParts[++oldStartIndex];
+        newStartResult = newResults[++newStartIndex];
+      } else if (oldEndPart.key == newEndResult.key) {
+        // Old tail matches new tail; update in place
+        newParts[newEndIndex] = updatePart(oldEndPart, newEndResult);
+        oldEndPart = oldParts[--oldEndIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (!newKeyToIndexMap.has(oldStartPart.key)) {
+        // Old head is no longer in new list; remove
+        removePart(oldStartPart);
+        oldStartPart = oldParts[++oldStartIndex];
+      } else if (!newKeyToIndexMap.has(oldEndPart.key)) {
+        // Old tail is no longer in new list; remove
+        removePart(oldEndPart);
+        oldEndPart = oldParts[--oldEndIndex];
+      } else if (oldStartPart.key == newEndResult.key) {
+        // Old head matches new tail; update and move to new tail
+        newParts[newEndIndex] = updatePart(oldStartPart, newEndResult);
+        movePart(directivePart, oldStartPart, newParts[newEndIndex + 1]);
+        oldStartPart = oldParts[++oldStartIndex];
+        newEndResult = newResults[--newEndIndex];
+      } else if (oldEndPart.key == newStartResult.key) {
+        // Old tail matches new head; update and move to new head
+        newParts[newStartIndex] = updatePart(oldEndPart, newStartResult);
+        movePart(directivePart, oldEndPart, oldStartPart);
```

Comment:
It's unclear why you move to before the `oldStartPart` and not before `newParts[newStartIndex]`?

Reply:
Yeah this is pretty tricky, you really have to draw a picture out to fully grok it.  But it comes down to this: the head insertBefore reference is always `oldStartPart`, and the tail insertBefore reference is always `newParts[newEndIndex+1]`.

Why? Because on the head side, we don't yet have `newParts[newStartIndex]` to insert before (that's the one we're currently working on, so the array is still empty there), however `oldStartPart` has not been moved yet, so it's still in the "next" position in the DOM for the head pointer.  On the tail side, `oldEndPart[oldEndIndex+1]` would be right except it may have already been moved; otherwise that would be what we'd go in front of; however we _do_ have a `newParts[newEndIndex+1]` which is the last correctly positioned item in the DOM on the tail side, since we've already moved left of that, so that's the thing to insert before.

The seeming asymmetry is an artifact of the pointers moving from opposite directions (and `newParts` filling in from the outside in), and the fact that insertBefore requires current+1 semantics for the reference.  If we adjusted the API to have an `insertAfter` option, it could look more balanced and be `inserAfter(..., newParts[newStartIndex-1])` and `insertBefore(..., newParts[newEndIndex+1])`.

Alternatively, we could just add a comment  stating something like this, since don't know how to succinctly say what I did above in comments:
> Note that as the head and tail pointers move inwards, the insertBefore reference is always `oldStartPart` for the head and `newParts[newEndIndex+1]` for the tail.

Reply:
Done, I added a more concise note to this effect.